crates/pruning/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+use citrea_evm::Evm;
+use futures::future;
+use serde::{Deserialize, Serialize};
+use sov_db::ledger_db::SharedLedgerOps;
+use sov_modules_api::default_context::DefaultContext;
+use tokio::select;
+use tokio::sync::broadcast;
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, info};
+
+#[cfg(test)]
+mod tests;
+
+/// A configuration type to define the behaviour of the pruner.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct PruningConfig {
+    /// Defines the number of blocks from the tip of the chain to remove.
+    pub distance: u64,
+}
+
+impl Default for PruningConfig {
+    fn default() -> Self {
+        Self { distance: 256 }
+    }
+}
+
+pub struct Pruner<DB>
+where
+    DB: SharedLedgerOps,
+{
+    /// config
+    config: PruningConfig,
+    /// The last block number which was pruned.
+    last_pruned_block: u64,
+    /// A channel receiver which gets notified of new L2 blocks.
+    receiver: broadcast::Receiver<u64>,
+    /// Access to ledger tables.
+    ledger_db: DB,
+}
+
+impl<DB> Pruner<DB>
+where
+    DB: SharedLedgerOps + Send + Sync + Clone + 'static,
+{
+    pub fn new(
+        config: PruningConfig,
+        last_pruned_block: u64,
+        receiver: broadcast::Receiver<u64>,
+        ledger_db: DB,
+    ) -> Self {
+        Self {
+            config,
+            last_pruned_block,
+            receiver,
+            ledger_db,
+        }
+    }
+
+    /// Prune everything
+    pub async fn prune(&self, up_to_block: u64) {
+        info!("Pruning up to L2 block: {}", up_to_block);
+        let ledger_db = self.ledger_db.clone();
+        let ledger_pruning_handle =
+            tokio::task::spawn_blocking(move || prune_ledger(ledger_db, up_to_block));
+        let evm_pruning_handle = tokio::task::spawn_blocking(move || prune_evm(up_to_block));
+
+        println!("Waiting for pruning task");
+        future::join_all([ledger_pruning_handle, evm_pruning_handle]).await;
+        println!("Finished pruning");
+    }
+
+    pub async fn run(mut self, cancellation_token: CancellationToken) {
+        loop {
+            select! {
+                biased;
+                _ = cancellation_token.cancelled() => {
+                    println!("Stopping pruner");
+                    // Store the last pruned l2 height in ledger DB to be restored in the next initialization.
```

Comment:
Is it required here? You set it when current_l2_block > distance, it should be enough

Reply:
I am only setting the ledger value when the pruning stops, to decrease the number of writes we do. When current_l2_block > distance, i am only setting a local member variable.

Reply:
Got it. what if the node dies with another signal, lets say `SIGABRT` instead of `SIGINT`?

Reply:
Yea good point. I'll attend to that.

Reply:
Apparently, `ctrl_c` handles `SIGINT` but `SIGABRT` would only happen if the task panics, in which case, tokio catches the panic and terminates the task as if it finished normally. We may need to handle `SIGTERM`.

Though this is out of scope of this PR IMO.

Reply:
Definitely out of scope of this pr. We definitely need to handle SIGTERM as it is the docker kill signal.